src/main/kotlin/pt/up/fe/ni/website/backend/service/RoleService.kt

```diff
@@ -1,46 +1,113 @@
+    private val activityService: ActivityService
 ) {
 
-    fun grantPermissionToRole(role: Role, permission: Permission) {
-        role.permissions.add(permission)
+    fun getRole(roleId: Long): Role {
+        val role = roleRepository.findById(roleId).orElseThrow {
+            throw NoSuchElementException(ErrorMessages.roleNotFound(roleId))
+        }
+        return role
+    }
+    fun getAllRoles(): List<Role> = roleRepository.findAll().toList()
+    fun grantPermissionToRole(roleId: Long, permissions: Permissions) {
+        val role = getRole(roleId)
+        role.permissions.addAll(permissions)
         roleRepository.save(role)
     }
 
-    fun revokePermissionFromRole(role: Role, permission: Permission) {
-        role.permissions.remove(permission)
+    fun revokePermissionFromRole(roleId: Long, permissions: Permissions) {
+        val role = getRole(roleId)
+        role.permissions.removeAll(permissions)
         roleRepository.save(role)
     }
 
-    fun grantPermissionToRoleOnActivity(role: Role, activity: Activity, permission: Permission) {
-        val foundActivity = activity.associatedRoles
+    fun grantPermissionToRoleOnActivity(roleId: Long, activityId: Long, permissions: Permissions) {
+        val activity = activityService.getActivityById(activityId)
+        val role = getRole(roleId)
+        val foundPerActivityRole = activity.associatedRoles
             .find { it.activity == activity } ?: PerActivityRole(Permissions())
-        foundActivity.role = role
-        foundActivity.activity = activity
+        foundPerActivityRole.role = role
+        foundPerActivityRole.activity = activity
 
-        foundActivity.role = role
-        foundActivity.permissions.add(permission)
-        perActivityRoleRepository.save(foundActivity)
+        foundPerActivityRole.permissions.addAll(permissions)
+        perActivityRoleRepository.save(foundPerActivityRole)
+        if (activity.associatedRoles.find { it.activity == activity } == null) {
+            role.associatedActivities.add(foundPerActivityRole)
+        }
+        // activityRepository.save(activity)
+        roleRepository.save(role)
     }
 
-    fun revokePermissionFromRoleOnActivity(role: Role, activity: Activity, permission: Permission) {
+    fun revokePermissionFromRoleOnActivity(roleId: Long, activityId: Long, permissions: Permissions) {
+        val activity = activityService.getActivityById(activityId)
+        val role = getRole(roleId)
         val foundActivity = activity.associatedRoles
             .find { it.role == role } ?: return
-
-        foundActivity.permissions.remove(permission)
+        foundActivity.permissions.removeAll(permissions)
         perActivityRoleRepository.save(foundActivity)
+        // activityRepository.save(activity)
+    }
+
+    fun createNewRole(dto: RoleDto): Role {
+        if (roleRepository.findByName(dto.name) != null) {
+            throw IllegalArgumentException(ErrorMessages.roleAlreadyExists)
+        }
```

Comment:
This isn't right. We can have roles with the same name as long as they're in different generations. I created a role `NoDuplicateRoles` In the generation model but it's only being validated when creating generations. Perhaps we can validate the generation again or port the logic somewhere else. What do you think?

Reply:
We can't use `@Valid` on expressions so I'm not sure how I would reuse `NoDuplicateRoles` in this case

Reply:
@LuisDuarte1 
- 1st option: Check `EntityDto`and see how we use the validator. We can use the same logic to validate the generation once the role is added. You can either validate before saving or let Spring rollback the changes when the exception is thrown (I think this is the behaviour since we're using @Transactional but please check)
- 2nd option: I don't see a good way tbh I don't like utils classes either but it's hard to find a better option than the 1st above. We could theoretically access the `isValid` method of `NoDuplicateRolesValidator` but that's cursed. The alternative would be a class with validation methods or repeating the 2 lines of code.

Imo the 1st option is cleaner

Reply:
I didn't test it but it seems possible to validate a single field if you want: https://codingexplained.com/coding/java/hibernate/validate-single-field-hibernate-validator

Reply:
Done ^^